checks.d/mapreduce.py

```diff
@@ -0,0 +1,500 @@
+                                                        'counter_name:' + str(counter_name).lower()]
+
+                                                self._set_metrics_from_json(tags,
+                                                    counter,
+                                                    MAPREDUCE_JOB_COUNTER_METRICS)
+
+    def _mapreduce_task_metrics(self, running_jobs, cluster_id):
+        '''
+        Get metrics for each MapReduce task
+        Return a dictionary of {task_id: 'tracking_url'} for each MapReduce task
+        '''
+        for job_id, job_stats in running_jobs.iteritems():
+
+            metrics_json = self._rest_request_to_json(job_stats['tracking_url'],
+                    'tasks',
+                    MAPREDUCE_SERVICE_CHECK)
+
+            if metrics_json.get('tasks'):
+                if metrics_json['tasks'].get('task'):
+
+                    for task in metrics_json['tasks']['task']:
+                        task_type = task.get('type')
+
+                        if task_type:
+                            tags = ['cluster_id:' + str(cluster_id),
+                                    'app_name:' + job_stats['app_name'],
+                                    'user_name:' + job_stats['user_name'],
+                                    'job_name:' + job_stats['job_name'],
+                                    'task_type:' + str(task_type).lower()
+                                    ]
+
+                            if task_type == 'MAP':
+                                self._set_metrics_from_json(tags, task, MAPREDUCE_MAP_TASK_METRICS)
+
+                            elif task_type == 'REDUCE':
+                                self._set_metrics_from_json(tags, task, MAPREDUCE_REDUCE_TASK_METRICS)
+
+    def _set_metrics_from_json(self, tags, metrics_json, metrics):
+        '''
+        Parse the JSON response and set the metrics
+        '''
+        for status, (metric_name, metric_type) in metrics.iteritems():
+            metric_status = metrics_json.get(status)
+
+            if metric_status is not None:
+                self._set_metric(metric_name,
+                    metric_type,
+                    metric_status,
+                    tags)
+
+    def _set_metric(self, metric_name, metric_type, value, tags=None, device_name=None):
+        '''
+        Set a metric
+        '''
+        if metric_type == GAUGE:
+            self.gauge(metric_name, value, tags=tags, device_name=device_name)
+        elif metric_type == HISTOGRAM:
+            self.histogram(metric_name, value, tags=tags, device_name=device_name)
+        else:
+            self.log.error('Metric type "%s" unknown' % (metric_type))
+
+    def _rest_request_to_json(self, address, object_path, service_name=SERVICE_CHECK, *args, **kwargs):
+        '''
+        Query the given URL and return the JSON response
+        '''
+        response_json = None
+
+        service_check_tags = ['resourcemanager_url:%s' % address]
```

Comment:
Just realized that this `address` wouldn't always be the same for a given `service_name`. Could we only use the `scheme://hostname:port` part of the `address` as a tag here? The tag also needs to be consistent with the tag that we send with the `OK` service check on the same `service_name`.

Also, given that it's not always the resource manager's url, we should probably simply name the tag `url`.